File: algo-method/41/algo41.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
#define rep(i, n) for(int i=0, i##_len=(n); i<i##_len; ++i)
#define all(x) (x).begin(),(x).end()

int main() {
    int N; cin >> N;
    // dp[i][t]: i日目にタスクtを選んだときの累計報酬
    vector<vector<int>> dp(N, vector<int>(3, 0));
    rep(i, N) {
        rep(t, 3) {
            int selected_before_reward = 0;
            int selected_before_task = -1;
            if (i > 0) {
                rep(before_task, 3) {
                    if (t == before_task) continue;
                    if (dp[i - 1][before_task] > selected_before_reward) {
                        selected_before_reward = dp[i - 1][before_task];
                        selected_before_task = before_task;
                    }
                }
            }
            if (selected_before_task == t) continue;
            int reward; cin >> reward;
            dp[i][t] = reward + selected_before_reward;
        }
    }
    cout << *max_element(all(dp[N - 1])) << endl;
}

File: algo-method/41/algo41.swift
func cinAsInt() -> Int { Int(readLine()!)! }
func cinAsIntList() -> [Int] { readLine()!.split(separator:" ").map{Int($0)!} }

let N = cinAsInt()
var dp: [[Int]] = Array.init(repeating: Array.init(repeating: 0, count: 3), count: N)

for i in 0..<N {
    let tasks = cinAsIntList()
    for t in 0...2 {
        var selectedBeforeReward = 0
        var selectedBeforeTask = -1
        if i > 0 {
            for beforeTask in 0...2 {
                guard t != beforeTask else { continue }
                if dp[i - 1][beforeTask] > selectedBeforeReward {
                    selectedBeforeReward = dp[i - 1][beforeTask]
                    selectedBeforeTask = beforeTask
                }
            }
        }
        guard selectedBeforeTask != t else { continue }
        let reward = tasks[t]
        dp[i][t] = reward + selectedBeforeReward
    }
}
print(dp[N - 1].max()!)
